pycode/memilio-epidata/memilio/ProgressIndicator.py

```diff
@@ -0,0 +1,282 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#############################################################################
+import memilio
+import sys
+import time
+import threading
+from os import get_terminal_size
+
+class ProgressIndicator:
+    """! Print an animation to show that something is happening.
+
+    Animations are rendered in a new thread, which is set up as deamon so that
+    it stops on main thread exit.
+    The methods Dots, Spinner and Percentage provide some default animations.
+    Start the animation by either using the `start`/`stop` functions or a
+    'with as' block, e.g
+    ```
+    import memilio.ProgressIndicator
+    with ProgressIndicator.Spinner():
+        do_something()
+    ```
+    or
+    ```
+    with ProgressIndicator.Percentage() as indicator:
+        for i in range(n) :
+            do_something()
+            indicator.set_progress((i+1)/n)
+    ```
+    """
+    def __init__(self, animator, delay):
+        """! Create an ProgressIndicator.
+
+        @param animator generator expression. The expression must loop, i.e.
+            it should never return StopIteration. `next(animator)` must be a
+            string of length < os.get_terminal_size().columns
+        @param delay positive real number. Sets delay in seconds between
+            drawing animation frames.
+        """
+        assert(delay > 0)
+        self._animator = animator
+        self._delay = delay
+        self._thread = None
+        self._enabled = False
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, exception, value, trace):
+        self.stop()
+
+    def _render(self):
+        """! Regularly update the animation. Do not call manually!"""
+        while self._enabled:
+            self.show()
+            # wait before writing next frame
+            time.sleep(self._delay)
+
+    def show(self):
+        """! Print and advance the animation. """
+        # write animation and proceed to next frame
+        sys.stdout.write(" {}\r".format(next(self._animator)))
+        sys.stdout.flush()
+
+    def start(self):
+        """! Start the animation in a new thread.
+
+        Must call stop() afterwards.
+        """
+        if not self._enabled:
+            self._enabled = True
+            # start new threat to render the animator in the background
+            self._thread = threading.Thread(target=self._render)
+            self._thread.setDaemon(True) # stops thread on main thread exit
+            self._thread.start()
+
+    def stop(self):
+        """! Stop the animation and join the thread. """
+        if self._enabled:
+            self._enabled = False
+            if self._thread and self._thread.is_alive():
+                self._thread.join()
+            sys.stdout.write("\033[K") # clear line
+
+def Spinner(delay=0.1, message=""):
```

Comment:
this (and `Dots`) are functions that are formatted as classes. Please be consistent and follow the python style guide. I would suggest Dots and Spinner should be classes like Percentage, i.e. containing a ProgressIndicator. Subclassing would also be fine.